clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp

```diff
@@ -0,0 +1,182 @@
+         (A.isStrictlyPositive() && B.isStrictlyPositive()) ||
+         (A.isZero() && B.isZero());
+}
+
+static bool greaterBySign(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNonNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isNonPositive());
+}
+
+bool TaggedUnionMemberCountCheck::isCountingEnumLikeName(
+    StringRef Name) const noexcept {
+  if (llvm::any_of(ParsedCountingEnumPrefixes,
+                   [&Name](const StringRef &prefix) -> bool {
+                     return Name.starts_with_insensitive(prefix);
+                   }))
+    return true;
+  if (llvm::any_of(ParsedCountingEnumSuffixes,
+                   [&Name](const StringRef &suffix) -> bool {
+                     return Name.ends_with_insensitive(suffix);
+                   }))
+    return true;
+  return false;
+}
+
+size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
+    const EnumDecl *Ed) const noexcept {
+  bool FoundMax = false;
+  llvm::APSInt MaxTagValue;
+  llvm::SmallSet<llvm::APSInt, 32> EnumValues;
+
+  // Heuristic for counter enum constants.
+  //
+  //   enum tag_with_counter {
+  //     tag1,
+  //     tag2,
+  //     tag_count, <-- Searching for these enum constants
+  //   };
+  //
+  // The final tag count is decreased by 1 if and only if:
+  // 1. There is only one counting enum constant,
+  // 2. The counting enum constant is the last enum constant that is defined,
+  // 3. The value of the counting enum constant is the largest out of every enum
+  //    constant.
+  // The 'ce' prefix is a shorthand for 'counting enum'.
+  size_t CeCount = 0;
+  bool IsLast = false;
+  llvm::APSInt CeValue = llvm::APSInt::get(0);
+
+  for (const auto &&[index, enumerator] : llvm::enumerate(Ed->enumerators())) {
+    const llvm::APSInt Val = enumerator->getInitVal();
+    EnumValues.insert(Val);
+    if (FoundMax) {
+      if (greaterBySign(Val, MaxTagValue) ||
+          (signEquals(Val, MaxTagValue) && Val > MaxTagValue)) {
+        MaxTagValue = Val;
+      }
+    } else {
+      MaxTagValue = Val;
+      FoundMax = true;
+    }
+
+    if (CountingEnumHeuristicIsEnabled) {
+      if (isCountingEnumLikeName(enumerator->getName())) {
+        IsLast = true;
+        CeValue = Val;
+        CeCount += 1;
+      } else {
+        IsLast = false;
+      }
+    }
+  }
+
+  size_t ValidValuesCount = EnumValues.size();
+  if (CeCount == 1 && IsLast && CeValue == MaxTagValue) {
+    ValidValuesCount -= 1;
+  }
+
+  return ValidValuesCount;
+}
+
+void TaggedUnionMemberCountCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *Root = Result.Nodes.getNodeAs<RecordDecl>("root");
+  const auto *UnionField = Result.Nodes.getNodeAs<FieldDecl>("union");
+  const auto *TagField = Result.Nodes.getNodeAs<FieldDecl>("tags");
+
+  // The matcher can only narrow down the type to recordType()
+  if (!isUnion(UnionField))
+    return;
+
+  if (hasMultipleUnionsOrEnums(Root))
+    return;
```

Comment:
The benefit of customised matchers with `AST_MATCHER` or `AST_MATCHER_P` instead of an imperative function call is that if we can sort out the non-matching cases declaratively, we spare ourselves the cost of building the `Result` object and firing the callback. (Also, there are *some* memoisation logics in the AST Matcher library, but that is only a theoretical benefit here, as this check only calls back each record once. The memoisation would benefit for checks where complex expressions are investigated.)